### PR TITLE
refactored category backend

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | yes
| Deprecations? | no
| Fixed tickets | fixes https://github.com/sulu/sulu/issues/355
| Related issues/PRs | https://github.com/massiveart/husky/pull/715
| License | MIT
| Documentation PR | -

#### What's in this PR?

This PR refactors the whole category backend (controller, manager, repository) as preperation for other category related improvements.

- Category api now enables patch requests (preperation for moving categories, #2691)
- Category api respects datagrid expand parameter to expand paths to children
- Fully tested CategoryController
- Cleaned up mxing of ApiEntity and DoctrineEntitiy in CategoryManager
- Using bundle specific Exceptions in CategoryManager now
- Cleaned up CategoryRepository
- The category tree in the frontend is lazy loaded now to avoid loading big trees on initialize
- Fixed category assignement for accounts

This PR is the first of three PRs regarding to the category backend refactoring.

#### BC Breaks/Deprecations

Several category related classes changed. Especially the CategoryRepository, CategoryManager and CategoryController were revised.
The final BC breaks are listed in https://github.com/sulu/sulu/pull/2862.

### CODE REVIEW

**Comment by nnatter** [comment]
@danrot: this PR should be ready for review. Unfortunately the tests failed because of a connection reset error in ContentBundle.

File: src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php [hunk 1]
```diff
@@ -235,246 +236,177 @@ public function getFieldDescriptors($locale)
...
-     * @param array $ids
-     *
-     * @return CategoryEntity[]
+     * {@inheritdoc}
      */
-    public function findByIds(array $ids)
+    public function find($locale, $parentId = null)
     {
-        return $this->categoryRepository->findCategoryByIds($ids);
+        if ($parentId && !$this->categoryRepository->findCategoryById($parentId)) {
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php]: Is it necessary to load the parent separately?

File: src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php [hunk 2]
```diff
@@ -235,246 +236,177 @@ public function getFieldDescriptors($locale)
...
+        }
+
+        $categoryEntity = $categoryWrapper->getEntity();
+        $this->em->persist($categoryEntity);
+
+        try {
+            $this->em->flush();
         } catch (UniqueConstraintViolationException $e) {
-            throw new KeyNotUniqueException($data['key'], $e);
+            throw new CategoryKeyNotUniqueException();
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php]: Wouldn't it be helpful to pass the non-unique key here?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 3]
```diff
@@ -114,75 +92,67 @@ public function getFieldsAction(Request $request)
...
+     * @param mixed $parentId
      *
      * @return \Symfony\Component\HttpFoundation\Response
      */
-    public function getChildrenAction(Request $request, $key)
+    public function getChildrenAction($parentId, Request $request)
     {
+        $locale = $this->getRequestParameter($request, 'locale', true);
+        // check if parent exists
+        $this->getManager()->findById($parentId, $locale);
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Please call this method `getCategoryManager`.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 4]
```diff
@@ -114,75 +92,67 @@ public function getFieldsAction(Request $request)
...
      *
      * @param Request $request
      *
      * @return \Symfony\Component\HttpFoundation\Response
      */
     public function cgetAction(Request $request)
     {
+        $locale = $this->getRequestParameter($request, 'locale', true);
+        $rootKey = $request->get('rootKey');
+        $rootId = ($rootKey) ? $this->getManager()->findByKey($rootKey, $locale)->getId() : null;
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Isn't that a `rootCategory`?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 5]
```diff
@@ -114,75 +92,67 @@ public function getFieldsAction(Request $request)
...
+
         if ($request->get('flat') == 'true') {
-            $list = $this->getCategoryListRepresentation($request);
+            $expandIds = array_filter(explode(',', $request->get('expand')));
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Why not call the parameters something like `expandId`?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 4 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Ah, have overseen the `getId` call... Is that necessary? Having a database call just to transfer a key to a id seems like a serious performance issue.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 6]
```diff
@@ -194,11 +164,11 @@ public function cgetAction(Request $request)
      */
     public function postAction(Request $request)
     {
-        return $this->saveEntity($request, null);
+        return $this->saveCategory($request);
     }
 
     /**
-     * Changes an existing category.
+     * Replaces the category which is assigned to the given id.
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: It doesn't really replace the category, it updates it, doesn't it?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 7]
```diff
@@ -207,21 +177,11 @@ public function postAction(Request $request)
...
-            $view = $this->view($exc->toArray(), 400);
-
-            return $this->handleView($view);
-        }
+        return $this->saveCategory($request, $id);
     }
 
     /**
-     * Partly changes an existing category.
+     * Updates the category which is assigned to the given id.
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Would keep the `partly` here, since it is what it really does.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 8]
```diff
@@ -230,93 +190,138 @@ public function putAction($id, Request $request)
...
 
-        $view = $this->responseDelete($id, $delete);
+        $view = $this->responseDelete($id, $deleteCallback);
 
         return $this->handleView($view);
     }
 
     /**
-     * @return RestHelperInterface
+     * Creates or updates an category based on the request.
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: `a category`

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 9]
```diff
@@ -230,93 +190,138 @@ public function putAction($id, Request $request)
...
-            $categoryWrapper = $categoryManager->getApiObject($categoryEntity, $locale);
-
-            $view = $this->view($categoryWrapper, 200);
-        } catch (EntityNotFoundException $e) {
-            $view = $this->view($e->toArray(), 404);
-        } catch (KeyNotUniqueException $e) {
-            $view = $this->view($e->toArray(), 400);
+        $listBuilder = $this->initializeListBuilder($locale);
+
+        if (!$request->get('search')) {
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: How is this connected to the actual content in the `if`?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 10]
```diff
@@ -230,93 +190,138 @@ public function putAction($id, Request $request)
...
-        $restHelper = $this->getRestHelper();
+        $restHelper = $this->get('sulu_core.doctrine_rest_helper');
 
         /** @var DoctrineListBuilderFactory $factory */
         $factory = $this->get('sulu_core.doctrine_list_builder_factory');
 
-        $listBuilder = $factory->create(self::$entityName);
-        $locale = $this->getRequestParameter($request, 'locale', true);
         $fieldDescriptors = $this->getManager()->getFieldDescriptors($locale);
-        $listBuilder->sort($fieldDescriptors['depth']);
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: How does sorting work without that? Is it disabled?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 11]
```diff
@@ -326,61 +331,85 @@ protected function getCategoryListRepresentation(Request $request, $parentKey =
...
+     * @param $rootIds array Ids of the categories of the root level
+     * @param $locale
+     *
+     * @return array|mixed
+     *
+     * @throws InvalidExpressionArgumentException
+     */
+    private function expandChildren($leafIds, $rootIds, $locale)
+    {
+        $leaves = $this->getManager()->findByIds($leafIds, $locale);
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: `$leaveCategories`, and is that really needed at all? Can't we load these categories in a more efficient way?

File: src/Sulu/Bundle/CategoryBundle/DependencyInjection/SuluCategoryExtension.php [hunk 12]
```diff
@@ -55,6 +58,9 @@ public function prepend(ContainerBuilder $container)
                 [
                     'exception' => [
                         'codes' => [
+                            CategoryNotFoundException::class => 404,
+                            CategoryKeyNotUniqueException::class => 409,
+                            CategoryNameMissingException::class => 409,
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/DependencyInjection/SuluCategoryExtension.php]: That's not a `409`, would rather say a `400`, because some data is missing, and not necessarily conflicting with anything else.

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php [hunk 13]
```diff
@@ -18,18 +18,11 @@
 
 /**
  * CategoryRepository.
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: Remove that and add a comment with sense

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php [hunk 14]
```diff
@@ -61,71 +54,44 @@ public function findCategoryByKey($key)
...
-            $queryBuilder = $this->getCategoryQuery();
-            $queryBuilder->andWhere('category.parent IS NULL');
-
-            if ($parent !== null) {
-                $queryBuilder->andWhere('categoryParent.id = :parentId');
-            }
-            if ($depth !== null) {
-                $queryBuilder->andWhere('category.depth = :depth');
+            if (empty($ids)) {
+                throw new NoResultException();
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: I would say returning an empty array here would be ok.

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php [hunk 15]
```diff
@@ -61,71 +54,44 @@ public function findCategoryByKey($key)
...
-     * @param number      $parent    the id of the parent to filter for
-     * @param number      $depth     the depth-level to filter for
-     * @param string|null $sortBy    column name to sort the categories by
-     * @param string|null $sortOrder sort order
-     *
-     * @return mixed|null
+     * {@inheritdoc}
      */
-    public function findCategories($parent = null, $depth = null, $sortBy = null, $sortOrder = null)
+    public function findCategoryByIds(array $ids)
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: `findCategoriesByIds`, since it is returning an id.

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php [hunk 16]
```diff
@@ -61,71 +54,44 @@ public function findCategoryByKey($key)
...
-     *
-     * @param int         $key       the key of the category to return the children for
-     * @param string|null $sortBy    column name to sort by
-     * @param string|null $sortOrder sort order
-     *
-     * @return Category[]
+     * {@inheritdoc}
      */
-    public function findChildren($key, $sortBy = null, $sortOrder = null)
+    public function findCategories($parentId = null)
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: `findCategoriesByParentId`

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Exception;
+
+class CategoryKeyNotUniqueException extends \Exception
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php]: Add the key as member variable

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Exception;
+
+class CategoryNotFoundException extends \Exception
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php]: Add the id as member variable and introduce a good error message here.

**Comment by danrot** [comment]
There are quite a lot database queries when loading the data with the `expand` flag, we should try to have a look if we can improve that.

File: src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php (hunk 1 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php]: @danrot: i am loading the parent to throw an explanitory exception, if the parent does not exist.
when we remove this line, the method will just return an empty list, if it is called with an invalid parent. would that be okay?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 4 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: i tried to implement the manager the repository as simple as possible. we could add another `find`-method to the manager and the repository with a `key`-parameter to prevent a second database call.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 5 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: are you talking about the api-parameter? sounds good for me, but i think we should use `expandId**s**`?

File: src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php (hunk 1 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php]: Maybe we can rewrite the entire query to get this working? I'll have quick look.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Yeah, you are right :)

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 4 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: That also wouldn't solve the issue from what I can see here... I think you'll need another `getListRepresentation` method, which takes a key as parameter.

Or another solution would be to not use the category key here at all. Why is this necessary? We used the id to get the children before also, didn't we?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 6 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: depends on the definition of replace/update. 
the put method sets all category-properties which are accesible by a get request. if you send a put request which only contains the id and the name, other accessible properties (like key) are deleted. in my opinion, an update would update the name but not remove the key (this is what an patch-request does).

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 7 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: same as above (putAction). 
but im fine with changing replace to update and update to partly update if we already use this terminology in other controllers.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 9 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: usually we are just returning the root-level of the request-respective category tree, but it would be unintuitive to just return the results in the root level, when answering a search request.
i will add a comment here.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 10 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: as we always respond with one level of the tree, we do not need to sortthe results by their tree-depth. this was a workaround for a bug in the frontend which is not capable of rendering children before their parents are rendered.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 11 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: see other `expand`-related comment.

**Comment by nnatter** [comment]
@danrot: thats true, im not perfectly happy with the solution either.
but its quite hard to find a simple solution which does not conflict pagination/sub-trees. we should talk about that.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 4 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: you are right..
there are category-lists in the frontend which are only displaying the subtree below a specified category (for example the category list in the edit view of the ContactBundle). These categories are specified in the configuration by their key (which makes sense as the id cannot set in the frontend).

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 7 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Yeah, we use it like that in other controllers. And it really doesn't replace it, since it only sets some values on the same object instance. Replace would mean for me that it creates a new instance.

Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: okay, i will fix that :)

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 10 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: And what about the `expand` parameter? This one is loading multiple levels, isn't it?

**Comment by danrot** [comment]
The category assignment in the does not work for the contacts anymore.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 10 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: yes, you are right. i just saw that this frontend bug was fixed with https://github.com/massiveart/husky/pull/451. so it should not be neccesary to sort by depth anymore.

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: This PR is already 1,5 years old, I am not sure if this still works correctly... We have included that line just a few weeks back. Please doublecheck if everything is working.

File: src/Sulu/Bundle/ContactBundle/DependencyInjection/Configuration.php [hunk 19]
```diff
@@ -64,7 +64,7 @@ public function getConfigTreeBuilder()
                         ->arrayNode('contact')
                             ->addDefaultsIfNotSet()
                             ->children()
-                                ->scalarNode('category_root')->defaultValue(null)->end()
+                                ->scalarNode('category_root')->defaultValue('test')->end()
```
Comment (nnatter) [inline src/Sulu/Bundle/ContactBundle/DependencyInjection/Configuration.php]: this should not be part of this pr :fearful:

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php (hunk 17 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php]: @danrot: done

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php (hunk 18 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php]: @danrot: done

**Comment by nnatter** [comment]
@danrot: The expandIds-flag should trigger constant 7 queries now (6 listbuilder queries + 1 query to determine paths). This PR should be ready for review again

File: src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php [hunk 20]
```diff
@@ -235,246 +236,187 @@ public function getFieldDescriptors($locale)
...
+            $categoryEntity->setCreator($user);
+        }
+
+        $categoryEntity->setChanger($user);
+        $categoryWrapper = $this->getApiObject($categoryEntity, $locale);
+
+        if (!$patch || $this->getProperty($data, 'name')) {
+            $categoryWrapper->setName($this->getProperty($data, 'name', null));
+        }
+        if (!$patch || $this->getProperty($data, 'key')) {
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php]: Would save `$this->getProperty($data, 'key')` in a separate variable, so that it is possible to reuse the value in the exception later.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 21]
```diff
@@ -230,114 +190,106 @@ public function putAction($id, Request $request)
...
-
-        /** @var DoctrineListBuilderFactory $factory */
-        $factory = $this->get('sulu_core.doctrine_list_builder_factory');
+        $listBuilder = $this->initializeListBuilder($locale);
 
-        $listBuilder = $factory->create(self::$entityName);
-        $locale = $this->getRequestParameter($request, 'locale', true);
-        $fieldDescriptors = $this->getManager()->getFieldDescriptors($locale);
-        $listBuilder->sort($fieldDescriptors['depth']);
+        // if search is set, return all matching categories. else return only level below parent in category tree
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: `categories, else`

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: and `return only one level below`

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 22]
```diff
@@ -230,114 +190,106 @@ public function putAction($id, Request $request)
...
-        if ($parentKey !== null) {
-            $this->addParentSelector($parentKey, $listBuilder);
+        if ($expandIds) {
+            // expand the paths to the categories which are assigned to the given expandIds.
+            // we need to do this in a second query, because we need to expand children of the current page only.
+            $resultIds = array_map(
+                function ($category) {
+                    return $category['id'];
+                },
+                $results
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Do you really need the result of the `ListBuilder` here? You know that the list builder adds some serious overhead (e.g. the second query for counting). Can't we solve this with a more efficient query?

Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Ok, there doesn't seem to be a second query for counting when only calling `execute`, but there might still be a more performant way to get these IDs, don't you think so?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php [hunk 23]
```diff
@@ -353,27 +305,73 @@ protected function getCategoryListRepresentation(Request $request, $parentKey =
...
+    private function expandChildren($leafIds, $rootIds, $locale)
+    {
+        // collect ids of categories between root level (inclusive) and leaves (exclusive)
+        $pathCategoryIds = [];
+        foreach ($this->getCategoryManager()->findBetween($rootIds, $leafIds, $locale) as $betweenCategory) {
+            $pathCategoryIds[] = $betweenCategory->getId();
+
+            // add category of root level to path category, if it is connected with a found path
+            if (in_array($betweenCategory->getParent()->getId(), $rootIds)) {
+                $pathCategoryIds[] = $betweenCategory->getParent()->getId();
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Don't you add some of the categories twice this way?

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+{
+    /**
+     * @var mixed
+     */
+    private $categoryIdentifier;
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php]: I think we go with `Id` in all the other parts of Sulu, can you change that with all references in this class?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 22 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Actually [this method](https://github.com/sulu/sulu/blob/236010711cc5cb1d0d128f4e1cee211b2283f807/src/Sulu/Component/Rest/ListBuilder/Doctrine/DoctrineListBuilder.php#L145) would already return what you need, but the list builder will still do a second query. Making this method public also doesn't seem like a good idea, because the list builder for another storage would maybe not need such a method. That's why I would introduce a separate method in the `CategoryRepository` here, would save us another query for every request.

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php [hunk 25]
```diff
@@ -11,151 +11,97 @@
...
+     * {@inheritdoc}
      */
-    public function findCategoryByIds(array $ids)
+    public function findCategoriesBetween($fromIds, $toIds)
     {
-        try {
-            if (empty($ids)) {
-                throw new NoResultException();
-            }
+        $queryBuilder = $this->getCategoryQuery();
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: This method adds also much information that probably isn't necessary for the place it is used.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 22 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: Yes, the $result array contains the current page of the root level. 
If no expandIds-flag is set, this is the result which is returned.
If an expandIds-flag is set, the expanded paths to the leaves are merged with these results.

It would be possible to load these results only if no expandIds-flag is set and use a more complex query if there are leaves to expand, but do not think that pays off as we are allready in O(1) time complexity.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 23 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: @danrot: Theoretically yes, but that should not matter as we use this array only in an `IN-statement` in our query.

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php (hunk 24 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php]: I changed this to Identifier, because the exception is also thrown if a category for a specific key is not found (CategoryManager::findByKey). Should we use another exception for that case?

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php (hunk 25 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: @danrot: Isn't it confusing if the category manager returns another "type" of categories with this method compared to any other method of the manager?

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 22 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Well, pagination isn't the most complicated thing we've ever built... And apart from that I am not sure if it really makes sense anyway, maybe we can remove it? /cc @chirimoya 

Apart from that we have a complexity of O(n), since the time is not fixed. I would still go with the separate query, and not using the listbuilder, which was not build for only retrieving the IDs of something and does massive overfetching in this case.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 23 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: Still not very clean, and I am not sure if every database has no problem with that. Can't this loop be rewritten to function correctly?

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php (hunk 24 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNotFoundException.php]: I would say so... Because it is not clear what the identifier is otherwise.

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php (hunk 25 above)
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryRepository.php]: This is the repository, not the manager. And yes, I would say that we should do it this way, because in this case we do not need most of the joins being done here.

File: src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php (hunk 22 above)
Comment (nnatter) [inline src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php]: the result of the first listbuilder instance is **not** used for the ids only. this result contains the current page/searchresult of the request. 
If an expandIds flag is set, we need to expand the paths to this ids additionally. This is done with a second listbuilder instance.  The second instance only loads the path to the ids, nothing is loaded twice. The two results are merged.

Using two different listbuilder instances is not the perfect way, but replacing the first listbuilder instance with a repository-method means reimplementing pagination.
Pagination in this case is quite complicated, as it is dependent of the sorting, which is again dependent on the translation.

**Comment by nnatter** [comment]
@danrot, @chirimoya: i disabled the pagination in the category controller and therefore got rid of one listbuilder instance. this pr should be ready for review again

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryIdNotFoundException.php [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Exception;
+
+class CategoryIdNotFoundException extends \Exception
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryIdNotFoundException.php]: phpdoc

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotFoundException.php [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Exception;
+
+class CategoryKeyNotFoundException extends \Exception
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotFoundException.php]: also here

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php (hunk 17 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryKeyNotUniqueException.php]: phpdoc (:

File: src/Sulu/Bundle/CategoryBundle/Exception/CategoryNameMissingException.php [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Exception;
+
+class CategoryNameMissingException extends \Exception
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Exception/CategoryNameMissingException.php]: (:

**Comment by chirimoya** [comment]
@nnatter ... I've reverted the merge. After the update we had a lot of issues. Not only the API of the Twig extensions changed, also the save a child category destroyed the tree. /cc @danrot
